File: scripts/05_clustering.py
import scanpy as sc
import pandas as pd
import logging
from pathlib import Path
import os

# -----------------------------
# Step 5: Clustering and Marker Gene Analysis
# -----------------------------

# -----------------------------
# User-adjustable parameters
# -----------------------------
LEIDEN_RESOLUTION = 0.5        # Adjust clustering granularity (higher = more clusters)
N_NEIGHBORS = 15               # Number of neighbors to use for graph construction (adjust based on data)
N_PCS = 40                     # Number of PCs to use
RANK_GENES_METHOD = "wilcoxon"  # Marker gene ranking method
TOP_N_GENES = 10               # Number of top marker genes per cluster to save in summary

# -----------------------------
# Setup
# -----------------------------
sc.settings.verbosity = 2
sc.settings.autoshow = False
sc.settings.figdir = "figures"
Path(sc.settings.figdir).mkdir(parents=True, exist_ok=True)

os.makedirs("results", exist_ok=True)
log_file = "results/05_log.txt"
logging.basicConfig(filename=log_file, level=logging.INFO, format="%(asctime)s %(message)s")
logging.info("Step 5 started: Clustering and Marker Gene Analysis")

logging.info(f"Parameters - LEIDEN_RESOLUTION: {LEIDEN_RESOLUTION}, N_NEIGHBORS: {N_NEIGHBORS}, "
             f"N_PCS: {N_PCS}, RANK_GENES_METHOD: {RANK_GENES_METHOD}, TOP_N_GENES: {TOP_N_GENES}")

# Load normalized and reduced data from Step 4
adata = sc.read("results/04_norm_dr_data.h5ad")
#print(adata.raw)  # Should not be None
#print(adata.raw.X[:5, :5])  # Check some values (should look log-transformed)
#print(adata.X[:5, :5])      # Check current data matrix values for comparison
logging.info(f"Loaded normalized data: {adata.n_obs} cells, {adata.n_vars} genes")

# -----------------------------
# Recompute neighbors if needed (adjust parameters)
# -----------------------------
logging.info(f"Computing neighbors with n_neighbors={N_NEIGHBORS}, n_pcs={N_PCS}")
sc.pp.neighbors(adata, n_neighbors=N_NEIGHBORS, n_pcs=N_PCS)

# -----------------------------
# Clustering with Leiden algorithm
# -----------------------------
logging.info(f"Running Leiden clustering with resolution={LEIDEN_RESOLUTION}")
sc.tl.leiden(adata, resolution=LEIDEN_RESOLUTION)

# Save clustering labels in obs
adata.obs["cluster"] = adata.obs["leiden"]

# Log cluster sizes
cluster_sizes = adata.obs['cluster'].value_counts().sort_index()
logging.info(f"Cluster sizes:\n{cluster_sizes.to_string()}")

# -----------------------------
# Plot UMAP with clusters and treatment metadata
# -----------------------------
logging.info("Generating UMAP plots colored by clusters and treatment")

# UMAP colored by cluster
sc.pl.umap(adata, color="cluster", save="_05_umap_clusters.png", show=False)

# UMAP colored by treatment (if present)
if "treatment" in adata.obs.columns:
    sc.pl.umap(adata, color="treatment", save="_05_umap_treatment.png", show=False)
else:
    logging.warning("Metadata 'treatment' not found in adata.obs, skipping treatment plot.")

# -----------------------------
# Marker Gene Identification
# -----------------------------
logging.info(f"Computing marker genes per cluster using {RANK_GENES_METHOD} method")
sc.tl.rank_genes_groups(adata, groupby="cluster", method=RANK_GENES_METHOD)

# Save marker gene results to CSV files
marker_dir = Path("results/markers")
marker_dir.mkdir(exist_ok=True)

clusters = adata.obs["cluster"].cat.categories if hasattr(adata.obs["cluster"], "cat") else sorted(adata.obs["cluster"].unique())
summary_rows = []

for cluster in clusters:
    df = sc.get.rank_genes_groups_df(adata, group=cluster)
    df["cluster"] = cluster
    df = df[["cluster", "names", "logfoldchanges", "pvals_adj"]]
    df.columns = ["cluster", "gene", "logfoldchanges", "pvals_adj"]
    df.to_csv(marker_dir / f"marker_genes_cluster_{cluster}.csv", index=False)
    logging.info(f"Saved marker genes for cluster {cluster} with {len(df)} genes")

    summary_rows.append(df.head(TOP_N_GENES))

# Save top marker genes summary
summary_df = pd.concat(summary_rows, ignore_index=True)
summary_df.to_csv("results/05_top_marker_genes_summary.csv", index=False)
logging.info(f"Saved top {TOP_N_GENES} marker genes summary for all clusters.")

# -----------------------------
# Generate HTML Report
# -----------------------------
report_path = Path("results/05_report.html")
figures = sorted(Path(sc.settings.figdir).glob("*05_*.png"))

with open(report_path, "w") as f:
    f.write("<html><head><title>Step 5: Clustering & Marker Gene Analysis</title>\n")
    f.write("""
    <style>
        body { font-family: Arial, sans-serif; padding: 20px; }
        h1, h2 { color: #2c3e50; }
        img { border: 1px solid #ccc; padding: 10px; margin-bottom: 20px; width: 700px; }
        a { text-decoration: none; color: #2980b9; }
    </style>
    </head><body>
    """)
    f.write("<h1>Step 5 Report: Leiden Clustering and Marker Genes</h1>\n")

    f.write("<h2>UMAP Plots</h2>\n")
    for fig in figures:
        f.write(f'<img src="../{fig}" alt="{fig.name}"><br>\n')

    f.write("<h2>Marker Genes per Cluster</h2>\n<ul>\n")
    for cluster in clusters:
        f.write(f'<li><a href="../results/markers/marker_genes_cluster_{cluster}.csv">Marker genes for cluster {cluster}</a></li>\n')
    f.write("</ul>\n")

    f.write(f'<h2>Summary</h2>\n<p><a href="../results/05_top_marker_genes_summary.csv">Top {TOP_N_GENES} marker genes summary (CSV)</a></p>\n')
    f.write("</body></html>")

logging.info(f"Generated HTML report at {report_path}")

# -----------------------------
# Save the annotated data
# -----------------------------
adata.write("results/05_clustered_data.h5ad")
logging.info("Saved clustered AnnData object to results/05_clustered_data.h5ad")

print("✅ Step 5 complete: Clustering and marker gene analysis done and saved.")

File: R_scripts/05_clustering.R
# -----------------------------
# Step 5: Clustering and Marker Gene Analysis
# -----------------------------

# Clear environment, graphics, and console
rm(list = ls())            # Remove all variables from the workspace
graphics.off()             # Close all open graphics devices
cat("\014")                # Clear the console (works in RStudio)

library(Seurat)
library(ggplot2)
library(dplyr)

source("R_scripts/pipeline_utils.R")

# -----------------------------
# Parameters and Paths
# -----------------------------
output_dir <- "results"
plot_dir <- "figures"
marker_dir <- file.path(output_dir, "markers")
input_file <- file.path(output_dir, "04_norm_dr_data.rds")
output_file <- file.path(output_dir, "05_clustered_data.rds")
report_file <- file.path(output_dir, "05_report.html")
log_file <- file.path(output_dir, "05_log.txt")

umap_cluster_file <- file.path(plot_dir, "05_umap_clusters.png")
umap_treatment_file <- file.path(plot_dir, "05_umap_treatment.png")

if (!dir.exists(output_dir)) dir.create(output_dir, recursive = TRUE)
if (!dir.exists(plot_dir)) dir.create(plot_dir, recursive = TRUE)
if (!dir.exists(marker_dir)) dir.create(marker_dir, recursive = TRUE)

params <- list(
  leiden_resolution = 0.5,
  n_neighbors = 15, 
  n_pcs = 40,
  top_n_genes = 10
)


log_msg("Step 5 started: Clustering and Marker Gene Analysis")

# -----------------------------
# Load Data
# -----------------------------
seurat_obj <- readRDS(input_file)
log_msg(sprintf("Loaded normalized data: %d cells × %d genes", ncol(seurat_obj), nrow(seurat_obj)))

# -----------------------------
# Neighbors and Clustering
# -----------------------------
# Compute the k-nearest neighbors graph using the specified number of PCs
seurat_obj <- FindNeighbors(seurat_obj, dims = 1:params$n_pcs, verbose = TRUE)

# Perform clustering (e.g. Leiden/Louvain) at the specified resolution
seurat_obj <- FindClusters(seurat_obj, resolution = params$leiden_resolution)
log_msg(sprintf("Identified clusters: %s", paste(levels(seurat_obj$seurat_clusters), collapse = ", ")))

# -----------------------------
# UMAP plots
# -----------------------------
# Compute UMAP embedding using the same PCs as used for clustering
seurat_obj <- RunUMAP(seurat_obj, dims = 1:params$n_pcs)

# Generate and save a UMAP plot colored by cluster labels
umap_cluster_plot <- DimPlot(seurat_obj, reduction = "umap", group.by = "seurat_clusters", label = TRUE) +
  ggtitle("UMAP: Clusters")
ggsave(umap_cluster_file, umap_cluster_plot, width = 7, height = 5)
log_msg("Saved UMAP plot colored by clusters")

# Generate UMAP plot colored by treatment if available in metadata
if ("treatment" %in% colnames(seurat_obj[[]])) {
  umap_treatment_plot <- DimPlot(seurat_obj, reduction = "umap", group.by = "treatment") +
    ggtitle("UMAP: Treatment")
  ggsave(umap_treatment_file, umap_treatment_plot, width = 7, height = 5)
  log_msg("Saved UMAP plot colored by treatment")
} else {
  log_msg("Metadata 'treatment' not found; skipping UMAP treatment plot")
}

# -----------------------------
# Marker Gene Identification
# -----------------------------
# Identify marker genes for each cluster
# only.pos = TRUE → only positive markers (upregulated genes)
# min.pct = 0.25 → only consider genes expressed in at least 25% of cells in the cluster
# logfc.threshold = 0.25 → only consider genes with log2FC > 0.25
markers <- FindAllMarkers(
  seurat_obj,
  only.pos = TRUE,
  min.pct = 0.25,
  logfc.threshold = 0.25
)

log_msg("Computed marker genes for each cluster")

# Save marker genes per cluster and collect top N markers for summary
clusters <- levels(seurat_obj$seurat_clusters)

summary_list <- list()
for (cl in clusters) {
  # Filter markers for the current cluster and select relevant columns
  df <- markers %>%
    filter(cluster == cl) %>%
    select(cluster, gene, logfc = avg_log2FC, p_adj = p_val_adj)
  
  # Save markers for this cluster as CSV
  out_file <- file.path(marker_dir, sprintf("marker_genes_cluster_%s.csv", cl))
  write.csv(df, out_file, row.names = FALSE)
  log_msg(sprintf("Saved marker genes for cluster %s (%d genes)", cl, nrow(df)))
  
  # Store top N markers for summary output
  summary_list[[cl]] <- head(df, params$top_n_genes)
}

# Combine top markers from all clusters into a single summary file
summary_df <- bind_rows(summary_list)

# Save summary file
write.csv(summary_df, "results/05_top_marker_genes_summary.csv", row.names = FALSE)
log_msg(sprintf("Saved top %d marker genes summary for all clusters", params$top_n_genes))


# -----------------------------
# HTML Report
# -----------------------------
# Initialize the HTML content as a character vector with header and report title
html <- c(
  # HTML header with title and embedded CSS styles for formatting
  "<html><head><title>Step 5: Clustering & Marker Gene Analysis</title>",
  "<style>",
  "body { font-family: Arial, sans-serif; padding: 20px; }",
  "h1, h2 { color: #2c3e50; }",
  "img { border: 1px solid #ccc; padding: 10px; margin-bottom: 20px; width: 700px; }",
  "a { text-decoration: none; color: #2980b9; }",
  "</style></head><body>",
  
  # Report title
  "<h1>Step 5 Report: Leiden Clustering and Marker Genes</h1>",
  
  # Section for UMAP plots
  "<h2>UMAP Plots</h2>",
  
  # Embed UMAP plot showing clusters
  sprintf('<img src="../%s/05_umap_clusters.png"><br>', plot_dir)
)

# Add UMAP plot colored by treatment if the metadata exists
if ("treatment" %in% colnames(seurat_obj[[]])) {
  html <- c(html, sprintf('<img src="../%s/05_umap_treatment.png"><br>', plot_dir))
}

# Start section listing marker gene CSV files
html <- c(html, "<h2>Marker Genes per Cluster</h2><ul>")

# Loop through clusters and add a link to each marker gene CSV
for (cl in clusters) {
  html <- c(html, sprintf(
    '<li><a href="../%s/marker_genes_cluster_%s.csv">Marker genes for cluster %s</a></li>',
    marker_dir, cl, cl
  ))
}

# Add summary link and close HTML document
html <- c(html,
          "</ul>",
          sprintf(
            '<h2>Summary</h2><p><a href="../results/05_top_marker_genes_summary.csv">Top %d marker genes summary (CSV)</a></p>',
            params$top_n_genes
          ),
          "</body></html>"
)

# Write HTML content to report file
writeLines(html, report_file)
log_msg(sprintf("Generated HTML report at %s", report_file))

# -----------------------------
# Save the clustered Seurat object 
# -----------------------------
saveRDS(seurat_obj, output_file)
log_msg(sprintf("Saved clustered Seurat object to %s", output_file))


cat("✅ Step 5 complete: Clustering and marker gene analysis done and saved.\n")
